internal/pkg/api/handleEnroll.go

```diff
@@ -146,6 +158,33 @@ func (et *EnrollerT) _enroll(
 	}
 
 	agentID := u.String()
+	// only delete existing agent if it never checked in
+	if agent.Id != "" && agent.LastCheckin == "" {
+		zlog.Debug().
+			Str("EnrollmentId", enrollmentID).
+			Str("AgentId", agent.Id).
+			Str("APIKeyID", agent.AccessAPIKeyID).
+			Msg("Invalidate old api key and remove existing agent with the same enrollment_id")
+		// invalidate previous api key
+		err := invalidateAPIKey(ctx, zlog, et.bulker, agent.AccessAPIKeyID)
+		if err != nil {
+			zlog.Error().Err(err).
+				Str("EnrollmentId", enrollmentID).
+				Str("AgentId", agent.Id).
+				Str("APIKeyID", agent.AccessAPIKeyID).
+				Msg("Error when trying to invalidate API key of old agent with enrollment id")
+			return nil, err
+		}
+		// delete existing agent to recreate with new api key
+		err = deleteAgent(ctx, zlog, et.bulker, agent.Id)
```

Comment:
Is this safe if two calls to this API are made concurrently with the same enrollment ID? I could see this happening if somebody clones a VM with a pre-generated enrollment ID.

Since enrollment is done from the command line and the API call is not per agent, there is nothing guaranteeing this is impossible.

Reply:
Do you mean two agents enrolling at the same time with the same enrollment ID and one being deleted and it shouldn't? 
I think the same enrollment ID shouldn't be used by 2 different agents. If this happens we should have a way on the agent to generate a new enrollment ID. 

The goal of this feature is really to solve the agent duplication on enroll failure, this will unblock the flakyness of scale tests.

Reply:
> I think the same enrollment ID shouldn't be used by 2 different agents.  If this happens we should have a way on the agent to generate a new enrollment ID.

It shouldn't happen but people do clone virtual machine filesystems, and there is nothing preventing this from happening it will probably will eventually.

What would `deleteAgent` return if the agent didn't exist or was already deleted? We know the API key can be invalided already without issue already.

Reply:
Actually I'm not sure it matters, because if there are two calls with the same enrollment ID the deleteAgent call here will remove one of them and we'd be left with only one agent. The user can hopefully detect this.

I'd only expect this to be a problem if someone clones a virtual machine image where the enrollment ID has already been created.